### PR TITLE
Ch update histeft

### PR DESCRIPTION
**Pull request by jlawrenc**
This change modifies the topeft_run2 analysis processor and supporting scripts. The key change in the channel definitions are moved to a json as a common source for the scripts to draw upon. It also includes two new run modes the onZ/offZ channels and tau channels. The TOP-22-006 run mode has been fully validated to the data and asimov limits. The tau and onZ/offZ run modes are still undergoing final validation

### CODE REVIEW

**Comment by bryates** [comment]
Thanks @jlawrenc. Is this branch in sync with the latest master updates? I see some merge conflicts.

**Comment by jlawrenc** [comment]
It does appear there are a couple merge conflicts we were unaware of. It looks like they are due to some recent tau changes so I will resolve them before a merge

File: analysis/topeft_run2/datacards_post_processing.py
Comment (bryates) [inline analysis/topeft_run2/datacards_post_processing.py:64]: Just to make sure, these names are now automatically built if you don't use any of the new flags?

**Comment by anpicci** [comment]
@bryates we have just created a PR in topcoffea to include an addition for the offZ analysis

File: analysis/topeft_run2/datacards_post_processing.py
Comment (jlawrenc) [reply analysis/topeft_run2/datacards_post_processing.py:64]: There are three flags for this script. -s for TOP-22-006, -z for onZ/offZ, and -t for taus. Any of these flags will create the list, but using none will not produce it and cause the script to fail.

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [inline analysis/topeft_run2/analysis_processor.py:877]: Why was this removed?

File: topeft/modules/axes.py
Comment (bryates) [inline topeft/modules/axes.py:72]: Should we consider adding the hadronic tau as the "l" in ptljmax? Maybe that's beyond the scope if this PR.

**Comment by bryates** [comment]
@jlawrenc how is this PR coming along?

**Comment by anpicci** [comment]
Hi @bryates, it looks like @jlawrenc is still getting some discrepancies in terms of TOP-22-006 fit results with his setup (which has not been pushed to this branch, btw), so the tau leg of the PR is not ready yet. One option would be to merge this PR as it is (once the conflicts are solved) and push the tau-related updates in a different PR. This would allow @ywan2 to moving forward with the implementation of the forward regions with this new categorization strategy. What do you think?

**Comment by bryates** [comment]
> Hi @bryates, it looks like @jlawrenc is still getting some discrepancies in terms of TOP-22-006 fit results with his setup (which has not been pushed to this branch, btw), so the tau leg of the PR is not ready yet. One option would be to merge this PR as it is (once the conflicts are solved) and push the tau-related updates in a different PR. This would allow @ywan2 to moving forward with the implementation of the forward regions with this new categorization strategy. What do you think?

I'm fine with this. Do you agree @Andrew42 and @sscruz?

**Comment by anpicci** [comment]
@bryates I think once we merge https://github.com/TopEFT/topeft/pull/450, we can align this PR to master and eventually merge it without the tau channels. What do you think?

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py]: @jlawrenc @ywan2 did you ever think of more descriptive names for these?

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py]: @ywan2 any thoughts on this?

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py]: @ywan2 @jlawrenc

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py:767]: @jlawrenc @ywan2

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py:898]: @jlawrenc

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py:877]: @jlawrenc @ywan2

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py]: @jlawrenc @ywan2

File: topeft/modules/axes.py
Comment (bryates) [reply topeft/modules/axes.py:72]: @jlawrenc

**Comment by bryates** [comment]
> @bryates I think once we merge #450, we can align this PR to master and eventually merge it without the tau channels. What do you think?

Sounds like a plan. We should consider raising an exception or warning for the tau channel. There's also a few unaddressed comments in this PR.

File: topeft/modules/object_selection.py
Comment (bryates) [reply topeft/modules/object_selection.py]: @jlawrenc

File: topeft/modules/object_selection.py
Comment (bryates) [reply topeft/modules/object_selection.py]: @jlawrenc @anpicci

**Comment by ywan2** [comment]
@jlawrenc, @anpicci, @bryates 
Would people like to merge this PR this Friday (Dec 20)before Christmas? I am trying to test out some changes with an extra command option including `--full-analysis` that combines both `2lssfwd` and `3loffZ`, if that works I will make a commit before Friday. Otherwise I will make a mild edit by Thursday night then we can go over this and merge on Friday. Whatever needs to be done can be separate thing after Christmas. What do people think?

**Comment by bryates** [comment]
> @jlawrenc, @anpicci, @bryates Would people like to merge this PR this Friday (Dec 20)before Christmas? I am trying to test out some changes with an extra command option including `--full-analysis` that combines both `2lssfwd` and `3loffZ`, if that works I will make a commit before Friday. Otherwise I will make a mild edit by Thursday night then we can go over this and merge on Friday. Whatever needs to be done can be separate thing after Christmas. What do people think?

Sounds like a good plan. We can fix the tau issue in a separate branch.

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py:877]: Let's add this back in @ywan2

**Comment by anpicci** [comment]
@jlawrenc @ywan2 @bryates I have resolved the conflicts between master and this branch

**Comment by bryates** [comment]
> @jlawrenc @ywan2 @bryates I have resolved the conflicts between master and this branch

It looks like that introduced a lot if lint errors and the processor fails.

**Comment by bryates** [comment]
> > @jlawrenc @ywan2 @bryates I have resolved the conflicts between master and this branch
> 
> It looks like that introduced a lot if lint errors and the processor fails.

```
ModuleNotFoundError: No module named 'topcoffea.modules.CorrectedJetsFactory'
```
I'm thinking the TopCoffea branch checkout was overwritten when the master branch was merged in.

**Review by bryates — CHANGES_REQUESTED** [review]
Let's get the CI working, and address the remaining open comments (or make issues for them).

File: analysis/topeft_run2/analysis_processor.py
Comment (anpicci) [reply analysis/topeft_run2/analysis_processor.py]: @bryates a new proposal is coming with the next commit

File: analysis/topeft_run2/analysis_processor.py
Comment (anpicci) [reply analysis/topeft_run2/analysis_processor.py]: @bryates now they are enclosed in a if/else statement at L745

File: analysis/topeft_run2/analysis_processor.py
Comment (anpicci) [reply analysis/topeft_run2/analysis_processor.py]: @bryates this change has already been implemented

File: topeft/modules/object_selection.py
Comment (anpicci) [reply topeft/modules/object_selection.py]: @bryates it has already been addressed, removing the reference to the specific version

File: topeft/modules/object_selection.py
Comment (anpicci) [reply topeft/modules/object_selection.py]: also this has been addressed in the same way

File: analysis/topeft_run2/analysis_processor.py
Comment (anpicci) [reply analysis/topeft_run2/analysis_processor.py]: with the next commit, an if/else statement for these lines are proposed

**Comment by anpicci** [comment]
@bryates now everything should be addressed from my side, @jlawrenc and @ywan2 are going to cross-check your comments. Thanks!

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py:894]: @anpicci can you add this back in?

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py:921]: @anpicci what do you think?

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [reply analysis/topeft_run2/analysis_processor.py:945]: @anpicci

File: topeft/modules/axes.py
Comment (bryates) [reply topeft/modules/axes.py:72]: https://github.com/TopEFT/topeft/issues/454

Comment (jlawrenc) [reply topeft/modules/axes.py:72]: For the lj0pt, the tau is included when the tau_h_analysis flag is inititiated, but that is done in the analysis processor https://github.com/TopEFT/topeft/blob/ch_update_histeft/analysis/topeft_run2/analysis_processor.py#L851-L854

Comment (bryates) [reply topeft/modules/axes.py:72]: Great! Sorry I missed that. What about ptZ? I see `ptz_wtau`, but does that get used anywhere currently?

Comment (jlawrenc) [reply topeft/modules/axes.py:72]: ptZ is untouched, so under the tau analysis, it specifically only uses e/mu leptons not taus. ptz_wtau is ptz with a tau lepton pair. Used as a kinematic fit varibale for some of the 2lss+1tau+onZ channels

Comment (bryates) [reply topeft/modules/axes.py:72]: Ok, I see it now in the datacard post processing script.

**Comment by ywan2** [comment]
There were some lines accidentally deleted for the scalings selection, I added it back. This branch should be ready to merge.  
For fwd analysis, 2lss and 2lssfwd are binned by `lt`, whereas 2lss_4t is still `lj0pt`. The limits are similar versus if we change 2lss_4t to `lt` mining, but let me know if there are objections. For 3loffZ analysis, I am still working on investigation on the 3loffZ fake factor statstics, but there are no issue with the fits. We can leave the discussion after break.

**Comment by bryates** [comment]
> There were some lines accidentally deleted for the scalings selection, I added it back. This branch should be ready to merge. For fwd analysis, 2lss and 2lssfwd are binned by `lt`, whereas 2lss_4t is still `lj0pt`. The limits are similar versus if we change 2lss_4t to `lt` mining, but let me know if there are objections. For 3loffZ analysis, I am still working on investigation on the 3loffZ fake factor statstics, but there are no issue with the fits. We can leave the discussion after break.

Great, thanks for fixing everything @ywan2 ! It looks like there are a few merge conflicts with the master branch related to the cfg files. Once those are fixed we should be able to merge.

File: analysis/topeft_run2/analysis_processor.py
Comment (bryates) [inline analysis/topeft_run2/analysis_processor.py]: @anpicci please remove

File: environment.yml
Comment (bryates) [inline environment.yml]: @anpicci please fix

**Comment by bryates** [comment]
@anpicci can you fix the new lint errors?

**Review by bryates — APPROVED** [review]

**Pull request merged**
